cornice/cors.py

```diff
@@ -87,23 +89,35 @@ def ensure_origin(service, request, response=None, **kwargs):
     response = response or request.response
 
     # Don't check this twice.
-    if not request.info.get('cors_checked', False):
-        method = _get_method(request)
+    if request.info.get('cors_checked', False):
+        return response
+
+    origin = request.headers.get('Origin')
+    if not origin:
+        always_cors = asbool(
+            request.registry.settings.get("cornice.always_cors")
+        )
+        # With this setting, if the service origins has "*", then
+        # always return CORS headers.
+        origins = getattr(service, "cors_origins", [])
+        if always_cors and "*" in origins:
+            origin = "*"
+        else:
+            return response
 
-        origin = request.headers.get('Origin')
-        if origin:
-            if not any([fnmatch.fnmatchcase(origin, o)
-                        for o in service.cors_origins_for(method)]):
-                request.errors.add('header', 'Origin',
-                                   '%s not allowed' % origin)
-            elif service.cors_support_credentials_for(method):
-                response.headers['Access-Control-Allow-Origin'] = origin
-            else:
-                if any([o == "*" for o in service.cors_origins_for(method)]):
-                    response.headers['Access-Control-Allow-Origin'] = '*'
-                else:
-                    response.headers['Access-Control-Allow-Origin'] = origin
-        request.info['cors_checked'] = True
+    method = _get_method(request)
+    if not any([fnmatch.fnmatchcase(origin, o)
+                for o in service.cors_origins_for(method)]):
+        request.errors.add('header', 'Origin',
+                           '%s not allowed' % origin)
+    elif service.cors_support_credentials_for(method):
+        response.headers['Access-Control-Allow-Origin'] = origin
+    else:
+        if any([o == "*" for o in service.cors_origins_for(method)]):
+            response.headers['Access-Control-Allow-Origin'] = '*'
```

Comment:
I'm considering if this test branch needs to be higher. It seems to me that you shouldn't allow both `*` and a hostname origin, but if you do, the `*` should be preferred to the value of the `Origin: ` request.

I'm installing the project to better understand the settings.

Reply:
I updated the test locally to check the value of the `Access-Control-Allow-Origin`, and added a second test that set the request header ``Origin: notmyidea.org``. I then tried some variants for `cors_origins` on the `bacon` `Service`. Here are the values of the ``Access-Control-Allow-Origin`` header:

|``cors_origins``|No ``Origin``|``Origin: notmyidea.org``|
|-:|:-|:-|
|``('*',)``|``*``|``*``|
|``('*', 'notmyidea.org')``|``*``|``*``|
|``('notmyidea.org',)``|_No header_|``notmyidea.org``|

So, my proposed corner case is not valid, and the docs appear to be correct.

However, it seems ``always_cors`` allows the ``Origin`` check to be skipped, which seems like it could be a problem for some configurations. Maybe when ``always_cors`` is set but ``*`` is not in ``cors_origins``, the code code should fall back to checking the origin?

Reply:
Good catch! I added some tests and changed the shape of the patch, which also makes the code paths clearer IMO :)

Thanks for being so thorough!